source/slides/04_troubleshooting_bash.rst

```diff
@@ -0,0 +1,237 @@
+* None of the reported links can be found on the site.
+
+Symptoms
+--------
+
+* Link growth pattern
+* Reported links sometimes have patterns, ``/path/page/2, /path/page/3, ...``
+
+Debugging
+---------
+
+* Change browser string to be the same as Google Bot
+* Nothing changes
+
+* Use ``curl`` to inspect page at a lower level, notice temporary redirects
+* Browsers follow redirect without reading page, bots read page.
+
+Cause
+-----
+
+* Every temporary redirect page was still generating page content
+
+  * Including links that were invalid (404!)
+
+Fix
+---
+
+* Changed webapp to use permanent redirect, not temporary
+* Wait for GA 404 reports to go back down.
+
+Bash
+====
+
+Basics
+------
+
+* Interpreted Language
+* Built to easily interact with a system, run other programs
+* Pipes!
+
+Useful Symbols
+--------------
+
+* ``|`` Pipe -- redirects stdout of left to stdin of right
+
+  * ``grep 'searchstring' files/* | less``
+
+* ``&&`` and ``||`` -- Logical AND and OR
+
+  * ``true || echo "never gets here"``
+  * ``false && echo 'runs if false fails, which it always does'``
+
+* ``2>`` and ``1>`` -- Redirect STDERR and STDOUT
+
+  * ``fping -g 10.0.0.0/24 2>&1 | grep unreachable`` Normally ``fping`` outputs
+    an error message for each unreachable host. ``2>&1`` redirects STDERR to STDOUT
+    so we can grep it
+
+More Useful Symbols
+-------------------
+
+* ``!$`` -- Last argument to last command
+
+  * ``cat /dir/; cd !$`` Did you use ``cat`` when you meant ``cd``? Easy fix!
+
+* ``for; do; done`` -- for loop
+
+  * ``for x in $(find . -type f); do echo $x is a file; done``
+
+* ``${var//}`` -- Delete text from var
+
+  * ``var="this is a var"; echo ${var//this is }``
+
+* ``$()`` and `````` -- Run this as another bash command, and insert its output in place
+
+  * ``ls -l `which bash```
+
+Combining These Together
+------------------------
+
+.. code-block:: bash
+
+    $ set -a blocks
+    $ blocks="10.0.0.0/24"
+    $ set -a ips
+    $ ips=`fping -g 10.0.0.0/24 2>&1 | grep unreachable | tr \\  \\n`
+    $ for ip in $ips; do nmap -p 22 $ip && ips=`echo ${ips//$ip} | tr -s \\n`
+    $ echo $ips
+
+Function Definitions
+--------------------
+
+.. code-block:: bash
+
+    name () {
+    # code goes here
+    }
+
+Internal Variables
+------------------
+
+You should know the following:
+
+.. csv-table::
+   :header: Variable,Meaning
+
+   ``$*``,All arguments passed
+   ``$?``,Return code of last command run
+   ``"$@"``,All arguments passed as a list
+   ``$CDPATH``,Colon-delimited list of places to look for dirs
+   ``$HOME``, Location of user homedir
+   ``$IFS``,Internal Field Seperator
+   ``$OLDPWD``,Previous PWD
+
+Internal Variables
+------------------
+
+.. csv-table::
+   :header: Variable,Meaning
+
+   ``$PATH``,Colon-delimited list of places to find executables
+   ``$PWD``,Present Working Directory
+   ``$SHELL``,Path to running shell
+   ``$UID``, User ID
+   ``$USER``,Username
```

Comment:
You might also mention looking over the ``EXPANSION`` section of the bash man page. Has lots of useful information about what you can do with bash variables.